### PR TITLE
test: run e2e tests in parallel

### PR DESCRIPTION
**Pull request by sobolk**
*Issue #, if available:*

*Description of changes:*

This PR enables parallelization in our e2e tests.

Runtime before:
image

Runtime after:
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/integration-tests/src/test-e2e/minimal_with_typescript_idioms.ts [hunk 1]
```diff
@@ -4,11 +4,46 @@ import assert from 'node:assert';
...
+  readonly name = 'typescript-idiom';
+
+  /**
+   * Creates project creator.
+   */
+  constructor(private readonly cfnClient: CloudFormationClient) {}
+
+  createProject = async (e2eProjectDir: string): Promise<TestProjectBase> => {
+    const { projectName, projectRoot, projectAmplifyDir } =
+      await createEmptyAmplifyProject('typescript-idiom', e2eProjectDir);
```
Comment (Amplifiyer) [inline packages/integration-tests/src/test-e2e/minimal_with_typescript_idioms.ts:24]: you can use the `this.name` here?

File: packages/integration-tests/src/test-e2e/data_storage_auth_with_triggers.ts
Comment (Amplifiyer) [inline packages/integration-tests/src/test-e2e/data_storage_auth_with_triggers.ts:1]: can we move these files in a `test-project-setup` directory or something? Keep the test files only here

File: packages/integration-tests/src/test-e2e/test_project.ts
Comment (Amplifiyer) [inline packages/integration-tests/src/test-e2e/test_project.ts:1]: This file no longer returns test_projects or contains, perhaps it should be renamed `test_project_creator_creator` 🤣  ?

**Review by Amplifiyer — DISMISSED** [review]

File: .github/actions/install_with_cache/action.yml [hunk 2]
```diff
@@ -10,7 +10,7 @@ runs:
         path: |
           node_modules
           packages/**/node_modules
-        key: ${{ runner.os }}-${{ hashFiles('package-lock.json') }}
+        key: ${{ runner.os }}-${{ hashFiles('package-lock.json') }}-v2
```
Comment (sobolk) [inline .github/actions/install_with_cache/action.yml:13]: This is needed to invalidate cache that's incompatible with larger worker and is leading to:
```
Prepare all required actions
Getting action download info
Download action repository 'actions/cache@885[2](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:2)2ab9f39a2ea568f7027eddc7d8d8bc9d59c8' (SHA:88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8)
Run ./.github/actions/restore_build_cache
Prepare all required actions
Getting action download info
Run ./.github/actions/restore_install_cache
Run actions/cache@88522ab9f[3](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:3)9a2ea568f7027eddc7d8d8bc9d59c8
Received 13[4](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:4)217728 of 24[5](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:6)825942 (54.[6](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:7)%), 126.2 MBs/sec
Received 245825942 of 245825942 (100.0%), 133.0 MBs/sec
Cache Size: ~234 MB (245825942 B)
"C:\Program Files\Git\usr\bin\tar.exe" -xf C:/a/_temp/00c0e1[7](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:8)2-e349-4f14-92f9-10005a622fb[8](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:10)/cache.tzst -P -C C:/a/amplify-backend/amplify-backend --force-local --use-compress-program "zstd -d"
Cache restored successfully
Cache restored from key: Windows-418f36fb2696ff545b34d4f90f527534293b1ff5110a4729ea71303b9b53485d
Run actions/cache@88522ab9f39a2ea568f7027eddc7d8d8bc9d59c8
Received 1342[17](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:20)728 of 167938[22](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:25)8 (79.9%), 114.7 MBs/sec
Received 167938228 of 167938228 (100.0%), 116.9 MBs/sec
Cache Size: ~160 MB (167938228 B)
"C:\Program Files\Git\usr\bin\tar.exe" -xf C:/a/_temp/bb0ebe96-47ed-43e9-890f-ead3ad[23](https://github.com/aws-amplify/amplify-backend/actions/runs/6899407583/job/18770983822#step:4:26)ea8b/cache.tzst -P -C C:/a/amplify-backend/amplify-backend --force-local --use-compress-program "zstd -d"
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib: Cannot mkdir: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/defaults.d.ts.map: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/string_maps.d.ts.map: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/construct.js: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/string_maps.js: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/trigger_events.d.ts: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/index.js: Cannot open: No medium found
/usr/bin/tar: node_modules/@aws-amplify/auth-construct-alpha/lib/trigger_events.js: Cannot open: No medium found
```

File: .github/workflows/health_checks.yml [hunk 3]
```diff
@@ -14,7 +14,10 @@ jobs:
   install:
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
+        # Windows install must happen on the same worker size as subsequent jobs.
+        # Larger workers use different drive (C: instead of D:) to check out project and NPM installation
+        # creates file system links that include drive letter.
+        os: [ubuntu-latest, macos-latest, amplify-backend_windows-latest_8-core]
```
Comment (sobolk) [inline .github/workflows/health_checks.yml:20]: See https://github.com/aws-amplify/amplify-backend/pull/674/files#r1396662876

File: .changeset/tall-turkeys-double.md [hunk 4]
```diff
@@ -0,0 +1,2 @@
+---
+---
```
Comment (alharris-at) [inline .changeset/tall-turkeys-double.md:2]: Looks like a duplicate changeset file?

**Review by alharris-at — APPROVED** [review]
Approving, since I don't think the dup changeset has any negative impact.

File: packages/integration-tests/src/test-e2e/deployment.test.ts [hunk 5]
```diff
@@ -17,155 +21,175 @@ import {
...
+            name: testBranch.branchName,
+            type: 'branch',
+          };
+        });
+
+        afterEach(async () => {
+          await testProject.tearDown(branchBackendIdentifier);
+        });
+
+        void it(`[${testProjectCreator.name}] deploys fully`, async () => {
```
Comment (sdstolworthy) [inline packages/integration-tests/src/test-e2e/deployment.test.ts:58]: nit: having project name here and on line 39 is repetitive

**Review by sdstolworthy — APPROVED** [review]

File: .changeset/tall-turkeys-double.md (hunk 4 above)
Comment (sobolk) [reply .changeset/tall-turkeys-double.md:2]: yeah, I had to add non-empty to force local publish.
This is harmless and disappears during releases.

**Review by Amplifiyer — APPROVED** [review]

File: packages/integration-tests/src/test-e2e/deployment.test.ts (hunk 5 above)
Comment (sobolk) [reply packages/integration-tests/src/test-e2e/deployment.test.ts:58]: same for sandbox, but it doesn't seem to produce bad output, so I skipped this.

File: packages/integration-tests/src/test-e2e/create_amplify.test.ts [hunk 6]
```diff
@@ -6,175 +6,206 @@ import * as os from 'os';
...
-            },
-            null,
-            2
-          )
-        );
+import { testConcurrencyLevel } from './test_concurrency.js';
+
+void describe(
+  'create-amplify script',
+  { concurrency: testConcurrencyLevel },
```
Comment (0618) [inline packages/integration-tests/src/test-e2e/create_amplify.test.ts:13]: Here's where the parallel magic happens. ref: https://nodejs.org/api/test.html#testname-options-fn

**Pull request merged**